src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -174,16 +172,106 @@ private sealed class RunningInstance
+                                s_scheduledStreamsCount--;
+
+                                if (s_scheduledStreamsCount == 0)
+                                {
+                                    // Stop the FS event message pump
+                                    Interop.RunLoop.CFRunLoopStop(s_watcherRunLoop);                                    
+                                    s_watcherRunLoop = IntPtr.Zero;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                private static void WatchForFileSystemEventsThreadStart(object args)
+                {
+                    var inputArgs = (object[])args;
+                    var runLoopStarted = (ManualResetEventSlim)inputArgs[0];
+                    var _eventStream = (SafeEventStreamHandle)inputArgs[1];
+                    // Get this thread's RunLoop
+                    var runLoop = Interop.RunLoop.CFRunLoopGetCurrent();
+                    s_watcherRunLoop = runLoop;
+                    Debug.Assert(s_watcherRunLoop != IntPtr.Zero);
+
+                    // Retain the RunLoop so that it doesn't get moved or cleaned up before we're done with it.
+                    IntPtr retainResult = Interop.CoreFoundation.CFRetain(runLoop);
+                    Debug.Assert(retainResult == runLoop, "CFRetain is supposed to return the input value");
+
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(_eventStream, runLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+
+                    runLoopStarted.Set();
+                    try
+                    {
+                        // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                        Interop.RunLoop.CFRunLoopRun();
+                    }
+                    finally
+                    {
+                        lock(s_lockObject)
+                        {                    
+                            Interop.CoreFoundation.CFRelease(runLoop);                                  
+                            s_watcherRunLoop = IntPtr.Zero;
+                        }
+                    }
+                }
+            }
+
+            private void CancellationCallback()
+            {
+                if (!_stopping && _eventStream != null)
+                {
+                    _stopping = true;
+
+                    try
+                    {
+                        // When we get here, we've requested to stop so cleanup the EventStream and unschedule from the RunLoop
+                        Interop.EventStream.FSEventStreamStop(_eventStream);
+                    }
+                    finally 
+                    {
+                        StaticWatcherRunLoopManager.UnscheduleFromRunLoop(_eventStream);
```

Comment:
Do we have any tests that create multiple watchers, then stop one, and validate that the other continues working and raising events?  Before that was less important as the implementations were unrelated, but now multiple instances share a loop, so we should ensure that situation works correctly.

If we don't already, we should also have tests that force the loop to be restarted, e.g. create a bunch of watchers, stop them all, and then create another set of watchers and validate that they all get the correct events.

Reply:
I've added several tests to check multiple watchers

Reply:
Could you please take a look at the tests? Do they look good?